### PR TITLE
[Distributed] Support pre-shared ncclUniqueId rendezvous for weight transfer

### PR DESCRIPTION
**Pull request by dharak-cohere**
## Purpose

The dense NCCL weight-transfer engine only rendezvous via a TCPStore, which
requires torch on every rank — blocking torch-free trainers (e.g. JAX). This
adds an alternative: the trainer mints an `ncclUniqueId` and shares the 128
bytes (base64) with workers out of band; workers join NCCL directly from it, no
TCPStore. The two modes are mutually exclusive and validated at the API
boundary. Existing TCP behavior is unchanged.

## Changes

- `NCCLWeightTransferInitInfo` gains optional `nccl_unique_id_b64` (kept out of
  `repr`); exactly one rendezvous mode must be set.
- Additive `PyNcclCommunicator.from_unique_id_bytes(...)`; `__init__` untouched.
- Worker init selects the mode from the init info. UID mode is worker-side only;
  the dense trainer engine stays TCPStore-only.

## Test Plan

New `tests/distributed/test_weight_transfer_nccl_uid.py`: rendezvous-mode
validation (unit) + two-GPU integration — a worker receives a tensor over the
UID path from (a) a vLLM `PyNcclCommunicator` peer and (b) a torch-free peer
using only a cupy buffer + raw `ncclBroadcast`.

```bash
.venv/bin/python -m pytest tests/distributed/test_weight_transfer.py \
    tests/distributed/test_weight_transfer_nccl_uid.py -q
# 109 passed (4xH100)
```

```
AMD (ROCm/RCCL) — MI300X (gfx942): Ran the branch in rocm/vllm:rocm7.0.0_vllm_0.11.2_20251210 (ROCm 7.0.0, torch 2.9, RCCL 2.26.6). Confirmed the wrapper loads librccl.so.1 with 128-byte
  UIDs. 

  • test_weight_transfer_nccl_uid.py: 15 passed — incl. test_nccl_weight_transfer_between_processes_uid (PyNccl/RCCL UID path) and
    test_nccl_weight_transfer_torch_free_trainer (cupy H2D + raw ncclBroadcast/ncclAllReduce over RCCL). No hang in the store-less path.
  • test_weight_transfer.py: 94 passed, no regressions.
```

## Not a duplicate

Checked open PRs per `AGENTS.md`. Adjacent NCCL weight-transfer work is distinct:
[#51520](https://github.com/vllm-project/vllm/pull/51520) (`nccl_m2n` resharding)
and [#53751](https://github.com/vllm-project/vllm/pull/53751) (sparse checkpoint
coords) are different transports/semantics and still require torch on the
trainer; [#53541](https://github.com/vllm-project/vllm/pull/53541) adds
server-level E2E tests. No open PR adds a pre-shared `ncclUniqueId` rendezvous
(`from_unique_id_bytes`).

## Model evaluation

N/A — changes only the NCCL rendezvous/init path, not model computation or
output.

## AI assistance

Developed with AI assistance (Cursor); the submitter reviewed every line and ran
the tests above.

### CODE REVIEW

**Comment by rishitdholakia13** [comment]
LGTM ! Nice Work, have tested this across different devices of Nvidia and AMD ? Could  we also post the performance numbers ! would be helpful in getting the PR approved faster.

File: vllm/distributed/device_communicators/pynccl.py [hunk 1]
```diff
@@ -145,6 +149,68 @@ def __init__(
...
+        assert isinstance(device, torch.device)
+        self.device = device
+        with torch.accelerator.device_index(device.index):
+            self.comm = self.nccl.ncclCommInitRank(
+                self.world_size, self.unique_id, self.rank
+            )
+            stream = current_stream()
+            data = torch.zeros(1, device=device)
+            self.all_reduce(data)
+            stream.synchronize()
```
Comment (dongluw) [inline vllm/distributed/device_communicators/pynccl.py]: **1. Init-time warmup `all_reduce` is an undocumented handshake for the JAX path**

This copies `__init__`'s 1-element `all_reduce` after `ncclCommInitRank`. The torch-free test has to issue a matching `ncclAllReduce` or both ranks deadlock — that is the actual JAX contract, and it currently only lives in a test docstring.

A JAX trainer that mints a UID and immediately `ncclBroadcast`s will hang. Options, in order of preference:

- Skip warmup on the UID factory (the first broadcast is the warmup), or
- Add `warmup: bool = True` so the worker can be told not to, or
- Document it as a required handshake on `NCCLWeightTransferInitInfo` and `from_unique_id_bytes`, not only in the test.

File: vllm/distributed/weight_transfer/nccl_common.py [hunk 2]
```diff
@@ -8,37 +8,100 @@
...
+      unique id out of band and share it (over HTTP, etc.). All ranks must enter
+      init concurrently; there is no store barrier on this path.
+    """
 
-    master_address: str
-    master_port: int
     rank_offset: int
     world_size: int
+    master_address: str | None = None
+    master_port: int | None = None
```
Comment (dongluw) [inline vllm/distributed/weight_transfer/nccl_common.py]: **2. Positional constructor order changed, despite the commit saying it did not**

On main this was `master_address, master_port, rank_offset, world_size`. Here it is `rank_offset, world_size, master_address=None, master_port=None`.

Keyword and dict callers are fine (examples already use keywords). Positional callers silently swap meaning at runtime — dataclasses do not type-check. The new unit test then locks in the *new* order.

Reordering is justified (optional fields cannot precede required ones). Either drop the “positional API preserved” claim, or make the dataclass `kw_only=True` so old positional calls fail loudly rather than swapping fields.

File: vllm/distributed/device_communicators/pynccl.py [hunk 3]
```diff
@@ -145,6 +149,68 @@ def __init__(
...
+            del data
+        return self
```
Comment (dongluw) [inline vllm/distributed/device_communicators/pynccl.py:214]: **3. `from_unique_id_bytes` duplicates `PyNcclCommunicator.__init__`**

Device normalization, `world_size == 1` / missing-lib disable, `ncclCommInitRank`, warmup — copied wholesale via `cls.__new__`. The next warmup or teardown change will only land in one path. Consider extracting a shared `_complete_init(unique_id, rank, world_size, device)` used by both.

File: vllm/distributed/device_communicators/pynccl.py [hunk 4]
```diff
@@ -145,6 +149,68 @@ def __init__(
...
+        if self.world_size == 1 or envs.VLLM_DISABLE_PYNCCL:
+            self.available = False
+            self.disabled = True
+            return self
+        try:
+            self.nccl = NCCLLibrary(library_path)
+        except Exception:
+            self.available = False
+            self.disabled = True
+            return self
```
Comment (dongluw) [inline vllm/distributed/device_communicators/pynccl.py]: **4. Silently disabling on missing NCCL is worse on this path**

Same `except Exception: available = False` as `__init__`. On TCPStore, a disabled worker fails in an obvious way. On UID, the trainer is already in `ncclCommInitRank` and will hang until timeout. Prefer raising when the caller explicitly asked to join from a UID.

File: vllm/distributed/weight_transfer/nccl_engine.py [hunk 5]
```diff
@@ -287,21 +294,27 @@ def trainer_init(
...
             packed=init_info.packed,
             packed_buffer_size_bytes=init_info.packed_buffer_size_bytes,
             packed_num_buffers=init_info.packed_num_buffers,
         )
+        # Ship only the TCP rendezvous; the worker defaults the unused UID field.
+        worker_init_info_dict = {
+            key: value
+            for key, value in asdict(worker_init_info).items()
+            if value is not None
+        }
```
Comment (dongluw) [inline vllm/distributed/weight_transfer/nccl_engine.py]: **5. `asdict(...); if value is not None` is duplicated and probably unnecessary**

Dense and sparse `trainer_init` both strip `None`. Validation is `has_uid = bool(nccl_unique_id_b64)`, so shipping `nccl_unique_id_b64: None` is already TCP mode. If you keep the filter, pull it into one helper so the two engines cannot drift.

File: vllm/distributed/weight_transfer/nccl_common.py [hunk 6]
```diff
@@ -8,37 +8,100 @@
...
+            raise ValueError(
+                f"{ctx}: need nccl_unique_id_b64, or both master_address and "
+                f"master_port (got master_address={master_address!r}, "
+                f"master_port={master_port!r})"
+            )
+        return None
+    try:
+        decoded = base64.b64decode(nccl_unique_id_b64, validate=True)
+    except (ValueError, TypeError) as e:
+        raise ValueError(f"{ctx}: nccl_unique_id_b64 is not valid base64") from e
```
Comment (dongluw) [inline vllm/distributed/weight_transfer/nccl_common.py]: **6. Base64 contract is under-specified**

`validate=True` means standard alphabet with padding, not urlsafe. 128 bytes → 172 chars with `==`. Worth one sentence on `NCCLWeightTransferInitInfo`: JAX `base64.urlsafe_b64encode` will be rejected.

**Review by dongluw — COMMENTED** [review]
Nice, well-scoped change — xor validation, UID kept out of `repr`, additive `from_unique_id_bytes`, and a real torch-free interop test. A few contract/API notes below, most importantly the init-time warmup `all_reduce` (comment 1), which is the actual handshake a JAX peer has to know about.

File: vllm/distributed/device_communicators/pynccl.py (hunk 1 above)
Comment (dharak-cohere) [reply vllm/distributed/device_communicators/pynccl.py]: good point i think all reduce is a good handshake to have for comm init. I added the requirement in the document

File: vllm/distributed/weight_transfer/nccl_common.py (hunk 2 above)
Comment (dharak-cohere) [reply vllm/distributed/weight_transfer/nccl_common.py]: Good catch, updated the mis-stated claim.

File: vllm/distributed/device_communicators/pynccl.py (hunk 3 above)
Comment (dharak-cohere) [reply vllm/distributed/device_communicators/pynccl.py:214]: This is intentional choice to prevent any changes to the main `__init__`

File: vllm/distributed/weight_transfer/nccl_common.py (hunk 6 above)
Comment (dharak-cohere) [reply vllm/distributed/weight_transfer/nccl_common.py]: good point, added into doc

**Comment by dharak-cohere** [comment]
> LGTM ! Nice Work, have tested this across different devices of Nvidia and AMD ? Could we also post the performance numbers ! would be helpful in getting the PR approved faster.

Thanks! Added AMD tests in the Test Plan summary. For perf I think the comm init is negligible compared to weight broadcast round and this change would have minimal impact

File: vllm/distributed/weight_transfer/nccl_common.py [hunk 7]
```diff
@@ -8,37 +8,118 @@
...
+    ``ncclCommInitRank`` (see ``PyNcclCommunicator.from_unique_id_bytes``), so a
+    foreign peer must issue a matching one-element ``all_reduce`` before any
+    other collective or all ranks deadlock.
+    """
 
-    master_address: str
-    master_port: int
     rank_offset: int
     world_size: int
+    master_address: str | None = None
```
Comment (aoshen02) [inline vllm/distributed/weight_transfer/nccl_common.py:95]: Let's better not change the position?

File: vllm/distributed/device_communicators/pynccl.py [hunk 8]
```diff
@@ -145,6 +149,77 @@ def __init__(
...
+        self.nccl_version = self.nccl.ncclGetRawVersion()
+        self.unique_id = self.nccl.unique_id_from_bytes(unique_id_bytes)
+
+        if isinstance(device, int):
+            device = torch.device(f"cuda:{device}")
+        elif isinstance(device, str):
+            device = torch.device(device)
+        assert isinstance(device, torch.device)
+        self.device = device
+        with torch.accelerator.device_index(device.index):
```
Comment (aoshen02) [inline vllm/distributed/device_communicators/pynccl.py]: Can we share some logic between from_unique_id_bytes and init?

File: vllm/distributed/weight_transfer/nccl_common.py (hunk 7 above)
Comment (dharak-cohere) [reply vllm/distributed/weight_transfer/nccl_common.py:95]: sure, made it `kw_only` as need to have required fields follow optional

File: vllm/distributed/device_communicators/pynccl.py (hunk 8 above)
Comment (dharak-cohere) [reply vllm/distributed/device_communicators/pynccl.py]: Sure, refactored out.

**Review by jeejeelee — APPROVED** [review]
